### ISSUE TITLE
Add include keyword/annotation in GDScript

### ISSUE THREAD

**Issue by me2beats** [godotengine/godot-proposals#1751]
**Describe the project you are working on:**
gdscript plugins

**Describe the problem or limitation you are having in your project:**

I very often want to split the behavior of a script into several behaviors, for example pick out some of them, so that I can reuse these behaviors later in other classes.

Usually if one of these behaviors can be implemented with one method that will be used in other scripts, I just put it in some utility class as a method/static function.

But it gets much more complicated if I want this behavior to work with (override) virtual methods, which often leads to a big confusing tangle of signal connections and even work-rounds.
in this case, I often have to sacrifice either reuse or clarity of the code.

The Godot inheritance system, which can only inherit **one** class/script (per level), does not help very well in such cases.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
The idea is to have an `include` keyword that would declare/define script properties and methods, the code of which would be **executed before** the code of the corresponding script methods in any case.

This will improve the usability of many reusable components.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Let's create some script-behavior that we want to reuse in other scripts
let's name it `a.gd`:

```gdscript
func _ready():
    print('hello')
```
this is just a simple regular script, which, when attached to a scene node, will print "hello" when the game starts.

Now let's create `b.gd` with one line of code:
```gdscript
include a.gd
```
What should it do (when the game starts)?
the same as a.gd (print 'hello')!

Now, what if `b.gd` already has overriding `_ready()`?
```gdscript
include a.gd

func _ready():
    print('world')
```

Then this `_ready` will be executed only after `a.gd` `_ready()`
That is, **included code should be executed in any case**.

**The same goes for properties**.

By the way, `a.gd` could include another script like
`a.gd`:
```gdscript
include x.gd

func _ready():
    print('hello')
```

`x.gd`:
```gdscript
func _ready():
    print('foo bar')
```
in this case `b.gd` should print:
```gdscript
foo bar
hello
world
```

and so on (**multilevel include**)

but **multiple include** would be even more useful:
`a.gd`:
```gdscript
func _ready():
    print('hello')
```
`b.gd`:
```gdscript
func _ready():
    print('world')
```
`c.gd`:
```gdscript
include a, b
```
this should print:
```
hello
world
```

the order is important so 
```gdscript
include b, a
```
would give `world hello`

as an **anotation** it could be look like:
```
@include c, a, b
```
or
```
@include (c, a, b)
```
```
@include
    c, a, b
```
```
@include
    c
    a
    b
```

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Most of the cases could be workarounded for example with `load()` or `connect()`+`emit_signal()`, 
but depending on the complexity of the code, such work-rounds could be quite confusing and poorly readable.
It would also force the user to create such an architecture and then make him depend on that architecture.

**Is there a reason why this should be core and not an add-on in the asset library?:**
GDScript feature

**Comment by pycbouh** [comment]
What you describe is how inheritance works in GDScript. You just need to explicitly call superclass’ methods to force the order of operation. The only limitation is in special methods like `_ready`.

That said, what you probably actually want is traits.

**Comment by hilfazer** [comment]
If a.gd inherits Foo and b.gd inherits Bar and includes a.gd which class would b.gd actually inherit? What if b.gd also includes c.gd (after including a.gd) that inherits Baz?

**Comment by me2beats** [comment]
@pycbouh Inheritance+superclasses unfortunately work only in some cases.
As well as traits.

**Comment by me2beats** [comment]
@hilfazer `b.gd` should inherit `Bar` if it extends `Bar`.
It shouldn't affect inheritance defined in scripts.
You can think of it as adding code to a script instead of copy-pasting it

**Comment by pycbouh** [comment]
> Inheritance+superclasses unfortunately work only in some cases.
> As well as traits.

As opposed to this proposal working in *all* cases? 🙃 You haven't provided a case that isn't solved now by inheritance and superclass methods (with the exception of special methods, but it's an engine limitation, not a GDScript limitation). And you haven't provided a case that cannot be solved by traits if they were available. In fact, you present this proposal as a way to share some partial code of a class between several classes, which is what traits are designed to solve (to the best of my knowledge of them).

Also, note that what hilfazer said highlights one of the many issues with multiple inheritance, which your proposal kind of hints towards. Godot provides great tools for composition which is the main alternative to inheritance. Code sharing can be done by nested nodes, or, in case of non-Node-derived classes, using standard practices like having an instance of a helper class as a property of the main class.

**Comment by me2beats** [comment]
Closing in favor of wiser solutions of the problem
